Fix deadlock between adding to queue and getting batch

csmarchbanks opened the pull request:
Do not block when trying to write a batch to the queue. This can cause
appends to lock forever if the only thing reading from the queue needs
the mutex to write. Instead, if batchQueue is full pop the sample that
was just added from the partial batch and return false. The code doing
the appending already handles retries with backoff.

Fixes: #10264 

cc: @beorn7 Do you think it is worth another patch in the v2.33 series?

I would like to write a failing test still but would appreciate some comments.

cstyan reviewed (commented):
Given that we already retry inside `enqueue` I think this is fine.

csmarchbanks commented on storage/remote/queue_manager.go:
```diff
@@ -995,26 +994,18 @@ func (s *shards) stop() {
 
-// enqueue data (sample or exemplar).  If we are currently in the process of shutting down or resharding,
-// will return false; in this case, you should back off and retry.
+// enqueue data (sample or exemplar). If the shard is full, shutting down, or
```
:+1: I added a sentence clarifying what it means for a shard to be full.

csmarchbanks commented on storage/remote/queue_manager.go:
```diff
@@ -995,26 +994,21 @@ func (s *shards) stop() {
-	s.writeMtx.Lock()
-	defer s.writeMtx.Unlock()
-
-	select {
-	case <-s.softShutdown:
-		return false
-	default:
-	}
```
This seems unnecessary to me since line 1014 also does a check for softShutdown? Not sure if I am missing something though.

beorn7 commented:
I'm happy to cut another 2.33 bugfix release with this.

@csmarchbanks: To facilitate that, could you change this PR to be on top of the release-2.33 branch?

csmarchbanks commented:
I apologize for the noise - note to self, do not change the base branch until AFTER rebasing...

csmarchbanks commented on storage/remote/queue_manager.go:
```diff
@@ -1053,6 +1045,11 @@ type sampleOrExemplar struct {
 
 func newQueue(batchSize, capacity int) *queue {
 	batches := capacity / batchSize
+	// Always create an unbuffered channel even if capacity is configured to be
+	// less than max_samples_per_send.
+	if batches == 0 {
+		batches = 1
+	}
```
This is the only new change after the rebase. It would probably fix the deadlock in most cases by itself (including the test), but I haven't convinced myself that it would fix all cases so I like the rest of the changes as well. The specific case I am still worried about is if enough samples to completely fill the queue happen to get appended right after the batch_send_deadline expires.

csmarchbanks commented:
Updated the test again just now, it will now also fail in the case where capacity > max_samples_per_send, reproducing both cases seen in the issue where a deadlock occurred in v2.33.3+.

I also updated the base backoff when retrying an Append, we now wait ~5ms to begin with (about the smallest amount of time a request can take in normal cases) and work up from there. That is to resume work more quickly after filling a queue/resharding.

cstyan commented on storage/remote/queue_manager_test.go:
```diff
@@ -413,13 +414,52 @@ func TestReshardPartialBatch(t *testing.T) {
 	m.Stop()
 }
 
+func TestQueueFilledDeadlock(t *testing.T) {
```
can we include a short comment of when this deadlock would have been hit without your code change? `a big scrape causing an append to instantly fill a queue at the same time a batch_send_deadline is hit`?

csmarchbanks replied:
:+1: Done.

cstyan reviewed (approved):
LGTM, did you want to include the pprof calls in the tests?

csmarchbanks commented:
I did intentionally leave the pprof calls in the tests. I found them to be very helpful to identify where the deadlock is, and figured that would be helpful for any future maintainers as well. Happy to remove them if you don't think they are worthwhile though.

Example output:
```
=== RUN   TestQueueFilledDeadlock
    queue_manager_test.go:459: Deadlock between sending and appending detected
goroutine profile: total 8
1 @ 0x43b656 0x40814c 0x407b78 0x5182f5 0x519f0e 0x5176a2 0x519dbf 0x518b1d 0xc2926b 0x43b287 0x46c1c1
#       0x5182f4        testing.(*T).Run+0x374          /usr/local/go/src/testing/testing.go:1307
#       0x519f0d        testing.runTests.func1+0x6d     /usr/local/go/src/testing/testing.go:1598
#       0x5176a1        testing.tRunner+0x101           /usr/local/go/src/testing/testing.go:1259
#       0x519dbe        testing.runTests+0x43e          /usr/local/go/src/testing/testing.go:1596
#       0x518b1c        testing.(*M).Run+0x51c          /usr/local/go/src/testing/testing.go:1504
#       0xc2926a        main.main+0x14a                 _testmain.go:157
#       0x43b286        runtime.main+0x226              /usr/local/go/src/runtime/proc.go:255

1 @ 0x43b656 0x44b352 0xb40b05 0x46c1c1
#       0xb40b04        github.com/prometheus/prometheus/tsdb/wal.(*Watcher).loop+0x304 /home/cmarchbanks/go/src/github.com/prometheus/prometheus/tsdb/wal/watcher.go:211

1 @ 0x43b656 0x44b352 0xbf79d2 0x46c1c1
#       0xbf79d1        github.com/prometheus/prometheus/storage/remote.(*MetadataWatcher).loop+0x111   /home/cmarchbanks/go/src/github.com/prometheus/prometheus/storage/remote/metadata_watcher.go:119

1 @ 0x43b656 0x44b352 0xbfe265 0x46c1c1
#       0xbfe264        github.com/prometheus/prometheus/storage/remote.(*QueueManager).updateShardsLoop+0x124  /home/cmarchbanks/go/src/github.com/prometheus/prometheus/storage/remote/queue_manager.go:768

1 @ 0x43b656 0x44b352 0xbff5a7 0x46c1c1
#       0xbff5a6        github.com/prometheus/prometheus/storage/remote.(*QueueManager).reshardLoop+0x106       /home/cmarchbanks/go/src/github.com/prometheus/prometheus/storage/remote/queue_manager.go:887

1 @ 0x43b656 0x44b352 0xc00829 0xc003b1 0xbfbec8 0xc14c85 0x46c1c1
#       0xc00828        github.com/prometheus/prometheus/storage/remote.(*queue).Append+0x1e8                   /home/cmarchbanks/go/src/github.com/prometheus/prometheus/storage/remote/queue_manager.go:1068
#       0xc003b0        github.com/prometheus/prometheus/storage/remote.(*shards).enqueue+0x1d0                 /home/cmarchbanks/go/src/github.com/prometheus/prometheus/storage/remote/queue_manager.go:1016
#       0xbfbec7        github.com/prometheus/prometheus/storage/remote.(*QueueManager).Append+0x487            /home/cmarchbanks/go/src/github.com/prometheus/prometheus/storage/remote/queue_manager.go:533
#       0xc14c84        github.com/prometheus/prometheus/storage/remote.TestQueueFilledDeadlock.func1+0x64      /home/cmarchbanks/go/src/github.com/prometheus/prometheus/storage/remote/queue_manager_test.go:453

1 @ 0x43b656 0x44c44c 0x44c426 0x4683c5 0x4753a5 0xc0148a 0xc01473 0x46c1c1
#       0x4683c4        sync.runtime_SemacquireMutex+0x24                                               /usr/local/go/src/runtime/sema.go:71
#       0x4753a4        sync.(*Mutex).lockSlow+0x164                                                    /usr/local/go/src/sync/mutex.go:138
#       0xc01489        sync.(*Mutex).Lock+0x389                                                        /usr/local/go/src/sync/mutex.go:81
#       0xc01472        github.com/prometheus/prometheus/storage/remote.(*shards).runShard+0x372        /home/cmarchbanks/go/src/github.com/prometheus/prometheus/storage/remote/queue_manager.go:1210

1 @ 0x4666e5 0x556315 0x55612d 0x5532ab 0xc14ba5 0x5176a2 0x46c1c1
#       0x4666e4        runtime/pprof.runtime_goroutineProfileWithLabels+0x24                           /usr/local/go/src/runtime/mprof.go:746
#       0x556314        runtime/pprof.writeRuntimeProfile+0xb4                                          /usr/local/go/src/runtime/pprof/pprof.go:724
#       0x55612c        runtime/pprof.writeGoroutine+0x4c                                               /usr/local/go/src/runtime/pprof/pprof.go:684
#       0x5532aa        runtime/pprof.(*Profile).WriteTo+0x14a                                          /usr/local/go/src/runtime/pprof/pprof.go:331
#       0xc14ba4        github.com/prometheus/prometheus/storage/remote.TestQueueFilledDeadlock+0x5a4   /home/cmarchbanks/go/src/github.com/prometheus/prometheus/storage/remote/queue_manager_test.go:460
#       0x5176a1        testing.tRunner+0x101                                                           /usr/local/go/src/testing/testing.go:1259

--- FAIL: TestQueueFilledDeadlock (2.05s)
FAIL
FAIL    github.com/prometheus/prometheus/storage/remote 2.058s
FAIL
```

cstyan commented:
:+1: I think for the failure cases on deadlock test cases they're worth keeping

beorn7 commented:
Great. Thank you. Will cut the bugfix release today.
